New DAGs menu (#3528)
astronomer/docs: astro/dags-overview.md

---
sidebar_label: 'Overview'
title: 'Write and run DAGs on Astro'
id: dags-overview
description: Learn about how developing DAGs to run on Astro is different from other platforms.
---

> Astro includes several features that enhance the Airflow development experience, from DAG creation to testing. Use this documentation to learn about all of the key differences between development on Astro and other platforms. 

Author: Rather than include that info here, I would instead link to this doc from our migration guides. It's more likely that a user will find this doc on their migration journey if it's linked from the migration documentation

> Astro includes several features that enhance the Airflow development experience, from DAG creation to testing. Use this documentation to learn about all of the key differences between development on Astro and other platforms. 

Author: I'm also not loving the second sentence, which makes it sound like a compliance burden to switch over to Astro development. I'm gonna try rewriting it to frame the migration experience more positively.

## Project structure

> To develop and run DAGs on Astronomer products, they must belong to an Astro project. An _Astro project_ contains all files required to run your DAGs both locally and on Astronomer. In addition to DAG files, an Astro project includes dependencies, your Astro Runtime version, and any other files your workers need access to when you run tasks.

Reviewer: ```suggestion
To develop and run DAGs on Astronomer products, your DAGs must belong to an Astro project. An _Astro project_ contains all files required to run your DAGs both locally and on Astronomer. In addition to DAG files, an Astro project includes dependencies, your Astro Runtime version, and any other files your workers need access to when you run tasks.
```
pronoun/antecedent nitpick

Your Astro Runtime version is defined in your Astro project Dockerfile. Unlike Apache Airflow, upgrading Astro Runtime does not require manually migrating your metadata database. To upgrade your version of Airflow, you change the Astro Runtime version listed in your project Dockerfile and rebuild your project. See [Upgrade Astro Runtime](upgrade-runtime.md).

Your Dockerfile is also where you can define additional runtime arguments that trigger whenever your project builds. You can use these arguments to mount resources, such as API tokens, to your Airflow environment without including the specific resources in your project files. See [Customize your Dockerfile](cli/customize-dockerfile.md) for more details.

## Testing environments

You can run your DAGs in several different environments across the Astro ecosystem. Each has its own purpose in the development lifecycle. 

> - **Local Airflow environment**: You can run DAGs on your local machine using the Astro CLI. This development method is most useful if you need to quickly iterate and test changes, such as when fixing a bug, or you're just getting started with Airflow. Testing locally is free, open source, and does not require an internet connection.

Author: Ah you know what, the Airflow UI doesn't need an internet connection per se but you DO need the internet to download the Runtime image for the first time.

> - **Deployment**: When you deploy your DAGs to Astro, they run on a managed Deployment. Use Deployments to run production code, or create a development Deployment to test changes over a longer period of time than you could in a local Airflow environment. To test code on a Deployment, you must have an [Astro account](log-in-to-astro.md), and an administrator on your team must grant you access to the Deployment. For more information about how to structure Deployments for specific development workflows, see [Connections and branch-based deploys](best-practices/connections-branch-deploys.md).

Author: This is more about where the DAGs run, rather than the act of deploying itself, hence the OG title

### Unit tests

Whether you're building your project locally or deploying to Astro, you can run unit tests with the Astro CLI to ensure that your code meets basic standards before you run your DAGs. See [Test your DAGs](https://docs.astronomer.io/astro/cli/test-your-astro-project-locally) for more information.

> ## DAG feature integrations

Author: Eh I disagree that these aren't integrations - I intentionally grouped features that exist on Airflow, but that we enhance in Astro. But not all of these are DAG-specific, so I changed that a bit

Astro includes several features that enhance existing Airflow features and integrate them with the Astro UI. 

- On Astro, the Astro UI renders [Airflow tags](https://airflow.apache.org/docs/apache-airflow/stable/howto/add-dag-tags.html) defined in your DAGs. Use tags to filter DAGs across all Deployments from a single screen.
- The [Astro Environment Manager](manage-connections-variables.md) allows you to manage Airflow connections directly from the Astro UI. Unlike with connections defined in the Airflow UI, you can apply connections from the Environment Manager across multiple Deployments and Workspaces, as well as in a local Airflow environment. 

> - Astro has built-in infrastructure to run the KubernetesPodOperator and Kubernetes executor, such as default Pod limits and requests. Resource limits and requests are set by default on Astro Deployments, meaning that tasks running in Kubernetes Pods never request more resources than expected. See [Run the Kubernetes executor](kubernetes-executor.md) and [Run the KubernetesPodOperator](kubernetespodoperator.md) for more specific instructions and examples. 

Author: Is the fact that Astro runs on Kubernetes the reason we are much more reliably running k8s-based tasks?

## DAG observability

In local Airflow environments, you can use the Airflow UI to check your DAG runs, task logs, and component logs just as you would in any other Airflow environment.

On Astro, in addition to the Airflow UI, you have access to the **DAGs** page. From here, you can manage DAG and task runs for any Deployment in your Workspace. See [View logs](view-logs.md) [Manage DAG runs](manage-dags.md) for more information.

> Astro includes additional features for alerting on specific DAG status, such as Astro alerts. There are some circumstances where Astronomer recommends configuring Astro alerts instead of Airflow SLAs or failure notifications. See [When to use Airflow or Astro alerts for your pipelines on Astro](airflow-vs-astro-alerts.md).

Author: ```suggestion
Astro includes additional features for alerting on specific DAG status, such as Astro alerts. There are some circumstances where Astronomer recommends configuring Astro alerts instead of Airflow SLAs or failure notifications. See [When to use Airflow or Astro alerts for your pipelines on Astro](best-practices/airflow-vs-astro-alerts.md).
```

## Testing environments

There are two ways to run DAGs within the Astro ecosystem: Either in a local Airflow environment or on a hosted Astro Deployment. Each has its own purpose in the development lifecycle.

- **Local Airflow environment**: You can run DAGs on your local machine using the Astro CLI. This development method is most useful if you need to quickly iterate and test changes, such as when fixing a bug, or you're just getting started with Airflow. Testing locally is free and open source.
- **Deployment**: When you deploy your DAGs to Astro, they run on a managed Deployment. Use Deployments to run production code, or create a development Deployment to test changes over a longer period of time than you could in a local Airflow environment. To test code on a Deployment, you must have an [Astro account](log-in-to-astro.md) and an administrator on your team must grant you access to the Deployment. For more information about how to structure Deployments for specific development workflows, see [Connections and branch-based deploys](best-practices/connections-branch-deploys.md).

### Unit tests

> Whether you're building your project locally or deploying to Astro, you can run unit tests with the Astro CLI to ensure that your code meets basic standards before you run your DAGs. See [Test your DAGs](https://docs.astronomer.io/astro/cli/test-your-astro-project-locally) for more information.

Author: ```suggestion
Whether you're building your project locally or deploying to Astro, you can run unit tests with the Astro CLI to ensure that your code meets basic standards before you run your DAGs. The Astro CLI includes a default set of unit tests which you can run alongside your own tests in a single testing sequence. See [Test your DAGs](https://docs.astronomer.io/astro/cli/test-your-astro-project-locally) for more information.
```

- The [Astro Environment Manager](manage-connections-variables.md) allows you to create and manage Airflow connections directly from the Astro UI. Instead of being limited to defining connections in the Airflow UI or with a secrets manager, you can create connections from the Environment Manager on Astro and use the connections in your local Airflow environment or across multiple Deployments and Workspaces.
- Astro has built-in infrastructure to run the KubernetesPodOperator and Kubernetes executor, such as default Pod limits and requests. Task-level resource limits and requests are set by default on Astro Deployments, which means that tasks running in Kubernetes Pods never request more resources than expected. See [Run the Kubernetes executor](kubernetes-executor.md) and [Run the KubernetesPodOperator](kubernetespodoperator.md) for more specific instructions and examples. 

## DAG observability

In local Airflow environments, you can use the Airflow UI to check your DAG runs, task logs, and component logs just as you would in any other Airflow environment.

> On Astro, you have access to the **DAGs** page in addition to the Airflow UI. From here, you can manage DAG and task runs for any Deployment in your Workspace. See [View logs](view-logs.md) [Manage DAG runs](manage-dags.md) for more information.

Author: ```suggestion
In the Astro UI, you have access to the **DAGs** page in addition to the Airflow UI. From here, you can manage DAG and task runs for any Deployment in your Workspace. See [View logs](view-logs.md) [Manage DAG runs](manage-dags.md) for more information.

Astronomer recommends using the UI that best fits the need your team. If you prefer managing your DAGs from a single place and find the way that the way the Astro page is designed helpful, you don't need to use the Airflow UI. If you're a longtime user of Apache Airflow, you might feel more comfortable in the Airflow UI and don't need to use the Astro UI.
```

## DAG Alerting

> Astro includes additional features, such as Astro alerts, for alerting on specific DAG statuses. There are some circumstances where Astronomer recommends configuring Astro alerts instead of Airflow SLAs or failure notifications because it can simplify your DAG code and make it easier to manage alerts across multiple DAGs. See [When to use Airflow or Astro alerts for your pipelines on Astro](best-practices/airflow-vs-astro-alerts.md).

Reviewer 2: ```suggestion
Astro supports a set of alerting features that in many cases replace Airflow SLAs or failure notifications that are supported in the open source project and that you would otherwise configure. There are some circumstances where Astronomer recommends configuring Astro alerts instead of Airflow SLAs or failure notifications because it can simplify your DAG code and make it easier to manage alerts across multiple DAGs. See [When to use Airflow or Astro alerts for your pipelines on Astro](best-practices/airflow-vs-astro-alerts.md).
```

You can potentially find better wording here but I would add a half sentence that explicitly states that this might affect how you write your DAGs. I.e. If you use Astro alerts, you can likely DELETE a bunch of code and make your DAGs easier to read. We kind of allude to that but let's connect the dots for the reader more directly

## Astro open source packages

> Although Astronomer's open source packages are not built into Astro, Astronomer recommends implementing them when possible to simplify your Airflow pipelines. Tools like the Astro SDK include prebuilt operators that eliminate boiler plate code from your Airflow DAGs.

Reviewer 2: This is confusing at first -- what open source packages are we talking about? I would add a half sentence introduction to the fact that Astro has these (which I guess is just 2) and name them.

Would also consider saying "See this section for more" maybe just to make it clear

> Although Astronomer's open source packages are not built into Astro, Astronomer recommends implementing them when possible to simplify your Airflow pipelines. Tools like the Astro SDK include prebuilt operators that eliminate boiler plate code from your Airflow DAGs.

Author: ```suggestion
In addition to its commercial products, Astronomer maintains several open source packages that enhance Apache Airflow. Although Astronomer's open source packages are not built into Astro, Astronomer recommends implementing them when possible to simplify your Airflow pipelines. Tools like the Astro SDK include prebuilt operators that eliminate boiler plate code from your Airflow DAGs.
```

- **Deployment**: When you deploy your DAGs to Astro, they run on a managed Deployment. Use Deployments to run production code, or create a development Deployment to test changes over a longer period of time than you could in a local Airflow environment. To test code on a Deployment, you must have an [Astro account](log-in-to-astro.md) and an administrator on your team must grant you access to the Deployment. For more information about how to structure Deployments for specific development workflows, see [Connections and branch-based deploys](best-practices/connections-branch-deploys.md).

### Unit tests

Whether you're building your project locally or deploying to Astro, you can run unit tests with the Astro CLI to ensure that your code meets basic standards before you run your DAGs. The Astro CLI includes a default set of unit tests which you can run alongside your own tests in a single sequence. See [Test your DAGs](https://docs.astronomer.io/astro/cli/test-your-astro-project-locally) for more information.

> ## Airflow feature integrations

Author: ```suggestion
## Airflow feature enhancements
```

- The [Astro Environment Manager](manage-connections-variables.md) allows you to create and manage Airflow connections directly from the Astro UI. Instead of being limited to defining connections in the Airflow UI or with a secrets manager, you can create connections from the Environment Manager on Astro and use the connections in your local Airflow environment or across multiple Deployments and Workspaces.
- Astro has built-in infrastructure to run the KubernetesPodOperator and Kubernetes executor, such as default Pod limits and requests. Task-level resource limits and requests are set by default on Astro Deployments, which means that tasks running in Kubernetes Pods never request more resources than expected. See [Run the Kubernetes executor](kubernetes-executor.md) and [Run the KubernetesPodOperator](kubernetespodoperator.md) for more specific instructions and examples. 

## DAG observability

In local Airflow environments, you can use the Airflow UI to check your DAG runs, task logs, and component logs just as you would in any other Airflow environment.

> In the Astro UI, you have access to the **DAGs** page in addition to the Airflow UI. From here, you can manage DAG and task runs for any Deployment in your Workspace. See [View logs](view-logs.md) [Manage DAG runs](manage-dags.md) for more information.

Author: ```suggestion
In the Astro UI, you have access to the **DAGs** page in addition to the Airflow UI. From here, you can manage DAG and task runs for any Deployment in your Workspace. 
```
